src/ucs/datastruct/lru.c

```diff
@@ -0,0 +1,145 @@
+*
+* See file LICENSE for terms.
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "khash.h"
+#include "list.h"
+#include "lru.h"
+#include "ucs/debug/memtrack_int.h"
+
+#include <stdint.h>
+
+
+typedef struct {
+    void           *key;
+    ucs_list_link_t list;
+} ucs_lru_element_t;
+
+
+__KHASH_TYPE(lru_hash, uint64_t, ucs_lru_element_t);
+
+
+__KHASH_IMPL(lru_hash, , uint64_t, ucs_lru_element_t, 1,
+             kh_int64_hash_func, kh_int64_hash_equal);
+
+
+typedef khash_t(lru_hash) ucs_lru_hash_t;
+
+/* LRU cache data structure */
+typedef struct ucs_lru {
+    /* Hash table of addresses as keys */
+    ucs_lru_hash_t  hash;
+    /* Linked list ordered by recent access */
+    ucs_list_link_t list;
+    /* Cache size */
+    size_t          size;
+    /* Cache capacity */
+    size_t          capacity;
+} ucs_lru_t;
+
+
+ucs_lru_h ucs_lru_init(size_t capacity)
+{
+    ucs_lru_h lru;
+
+    if (capacity == 0) {
+        return NULL;
+    }
+
+    lru = ucs_calloc(1, sizeof(ucs_lru_t), "ucs_lru");
+    if (lru == NULL) {
+        return NULL;
+    }
+
+    /* Init LRU hash table */
+    kh_init_inplace(lru_hash, &lru->hash);
+
+    /* Resize to get the required size. Need to allocate extra space for khash
+     * to be happy. */
+    if (kh_resize(lru_hash, &lru->hash, capacity * 2) < 0) {
+        ucs_free(lru);
+        return NULL;
+    }
+
+    /* Init LRU linked list */
+    ucs_list_head_init(&lru->list);
+
+    /* Init other fields */
+    lru->capacity = capacity;
+    lru->size     = 0;
+    return lru;
+}
+
+void ucs_lru_destroy(ucs_lru_h lru)
+{
+    kh_destroy_inplace(lru_hash, &lru->hash);
+    ucs_free(lru);
+}
+
+static void ucs_lru_pop(ucs_lru_h lru)
+{
+    ucs_list_del(lru->list.prev);
+}
+
+static void ucs_lru_push(ucs_lru_h lru, ucs_lru_element_t *elem)
+{
+    ucs_list_add_head(&lru->list, &elem->list);
+}
+
+static void ucs_lru_remove(ucs_lru_h lru)
+{
+    ucs_lru_element_t *tail;
+    khint_t iter;
+
+    tail = ucs_container_of(lru->list.prev, ucs_lru_element_t, list);
+    iter = kh_get(lru_hash, &lru->hash, (uint64_t)tail->key);
+
+    ucs_lru_pop(lru);
+    kh_del(lru_hash, &lru->hash, iter);
+}
+
+ucs_status_t ucs_lru_touch(ucs_lru_h lru, void *key)
+{
+    khint_t iter;
+    int ret;
+    ucs_lru_element_t *elem;
+
+    iter = kh_put(lru_hash, &lru->hash, (uint64_t)key, &ret);
+    if (ret == UCS_KH_PUT_FAILED) {
+        return UCS_ERR_NO_RESOURCE;
+    }
+
+    elem      = &kh_val(&lru->hash, iter);
+    elem->key = key;
+
+    if (ucs_likely(ret == UCS_KH_PUT_KEY_PRESENT)) {
+        ucs_list_del(&elem->list);
+    } else if (lru->size == lru->capacity) {
+        ucs_lru_remove(lru);
+    } else {
+        lru->size++;
+    }
+
+    ucs_lru_push(lru, elem);
+    return UCS_OK;
+}
+
+void ucs_lru_get(const ucs_lru_h lru, void **array, size_t *size_p)
+{
+    ucs_list_link_t *link;
+
+    /* Assume 'elements' has at least 'lru->capacity' elements */

```

Comment:
Is that a reasonable assumption for a "generic LRU"?

Reply:
actually I was referring to the "array" argument.